### DISCUSSION TITLE
Output switching depending on BLE RSSI?

### DISCUSSION THREAD

**Discussion by New-three** [arendst/Tasmota#21670] [Q&A]
I have installed Tasmota Bluettoth on my ESP32 mini. My SHelly button is displayed under Bluetooth. 
How can I switch an output depending on the RSSI value? 
The goal is when the Shelly button approaches the ESP32 mini, the output on which an LED strip is hanging is switched off

**Comment by barbudor** [comment]
I don't think this is a viable idea
The BLE RSSi is not updated often enough

If you still want to try it, you can most likely do it using rules or berry
But without more details on your configuration, I don't know what exact messages you have to suggest you a rule (many Mi32Options can change the format of message which impact how to write the rule)

**Reply by sfromis** [reply]
With a Button, you should not expect to have any radio transmission (thus RSSI), except when you push the button, to cause a signal to be sent, or *possibly* at long intervals.

An existing option is if you get a BLE beacon, which will transmit relatively often (maybe at something like 10-second intervals), and those transmissions can be picked up by Tasmota. May still be too slow for "approaches".

Of course, if the Shelly Button can act as a beacon, and you set up Tasmota to handle the messages, you have an option there.

**Reply by New-three** [reply]
I don't think this is a viable idea
The BLE RSSi is not updated often enough

If you still want to try it, you can most likely do it using rules or berry
But without more details on your configuration, I don't know what exact messages you have to suggest you a rule (many Mi32Options can change the format of message which impact how to write the rule)

do you need any more information?

**Comment by New-three** [comment]
the Shelly is set up as a beacon.
![image](https://github.com/arendst/Tasmota/assets/99345562/f4ff4867-2a8d-4264-b6f2-c3b5085b8866)

Console:
`18:17:19.648 RSL: STATE = {"Time":"2024-06-21T18:17:19","Uptime":"0T02:30:51","UptimeSec":9051,"Heap":42,"SleepMode":"Dynamic","Sleep":50,"LoadAvg":19,"MqttCount":0,"Berry":{"HeapUsed":21,"Objects":284},"Wifi":{"AP":1,"SSId":"2,4GHz RFID","BSSId":"44:4E:6D:23:7E:38","Channel":1,"Mode":"HT20","RSSI":72,"Signal":-64,"LinkCount":1,"Downtime":"0T00:00:04"}}
18:17:19.675 RSL: BLE = {"Time":"2024-06-21T18:17:19","BLEDevices":{"total":2,"F4B3B182966D":{"i":0,"n":"SBBT-002C","r":-30},"CCB11A37D49C":{"i":1,"n":"[TV] Samsung 5 Series (32)","r":-98}}}
18:17:19.688 RSL: BLE = {"Time":"2024-06-21T18:17:19","BLE":{"scans":444,"adverts":160479,"devices":2,"resets":0}}
18:22:19.681 RSL: STATE = {"Time":"2024-06-21T18:22:19","Uptime":"0T02:35:51","UptimeSec":9351,"Heap":42,"SleepMode":"Dynamic","Sleep":50,"LoadAvg":19,"MqttCount":0,"Berry":{"HeapUsed":21,"Objects":284},"Wifi":{"AP":1,"SSId":"2,4GHz RFID","BSSId":"44:4E:6D:23:7E:38","Channel":1,"Mode":"HT20","RSSI":68,"Signal":-66,"LinkCount":1,"Downtime":"0T00:00:04"}}
18:22:19.708 RSL: BLE = {"Time":"2024-06-21T18:22:19","BLEDevices":{"total":2,"F4B3B182966D":{"i":0,"n":"SBBT-002C","r":-34},"CCB11A37D49C":{"i":1,"n":"[TV] Samsung 5 Series (32)","r":-95}}}
18:22:19.721 RSL: BLE = {"Time":"2024-06-21T18:22:19","BLE":{"scans":459,"adverts":165701,"devices":2,"resets":0}}
18:27:19.643 RSL: STATE = {"Time":"2024-06-21T18:27:19","Uptime":"0T02:40:51","UptimeSec":9651,"Heap":40,"SleepMode":"Dynamic","Sleep":50,"LoadAvg":19,"MqttCount":0,"Berry":{"HeapUsed":21,"Objects":284},"Wifi":{"AP":1,"SSId":"2,4GHz RFID","BSSId":"44:4E:6D:23:7E:38","Channel":1,"Mode":"HT20","RSSI":68,"Signal":-66,"LinkCount":1,"Downtime":"0T00:00:04"}}
18:27:19.669 RSL: BLE = {"Time":"2024-06-21T18:27:19","BLEDevices":{"total":2,"F4B3B182966D":{"i":0,"n":"SBBT-002C","r":-35},"CCB11A37D49C":{"i":1,"n":"[TV] Samsung 5 Series (32)","r":-97}}}
18:27:19.683 RSL: BLE = {"Time":"2024-06-21T18:27:19","BLE":{"scans":474,"adverts":171092,"devices":2,"resets":0}}
18:32:19.685 RSL: STATE = {"Time":"2024-06-21T18:32:19","Uptime":"0T02:45:51","UptimeSec":9951,"Heap":42,"SleepMode":"Dynamic","Sleep":50,"LoadAvg":19,"MqttCount":0,"Berry":{"HeapUsed":21,"Objects":284},"Wifi":{"AP":1,"SSId":"2,4GHz RFID","BSSId":"44:4E:6D:23:7E:38","Channel":1,"Mode":"HT20","RSSI":68,"Signal":-66,"LinkCount":1,"Downtime":"0T00:00:04"}}
18:32:19.712 RSL: BLE = {"Time":"2024-06-21T18:32:19","BLEDevices":{"total":2,"F4B3B182966D":{"i":0,"n":"SBBT-002C","r":-37},"CCB11A37D49C":{"i":1,"n":"[TV] Samsung 5 Series (32)","r":-97}}}
18:32:19.724 RSL: BLE = {"Time":"2024-06-21T18:32:19","BLE":{"scans":489,"adverts":176356,"devices":2,"resets":0}}
18:37:19.647 RSL: STATE = {"Time":"2024-06-21T18:37:19","Uptime":"0T02:50:51","UptimeSec":10251,"Heap":39,"SleepMode":"Dynamic","Sleep":50,"LoadAvg":19,"MqttCount":0,"Berry":{"HeapUsed":21,"Objects":284},"Wifi":{"AP":1,"SSId":"2,4GHz RFID","BSSId":"44:4E:6D:23:7E:38","Channel":1,"Mode":"HT20","RSSI":68,"Signal":-66,"LinkCount":1,"Downtime":"0T00:00:04"}}
18:37:19.673 RSL: BLE = {"Time":"2024-06-21T18:37:19","BLEDevices":{"total":2,"F4B3B182966D":{"i":0,"n":"SBBT-002C","r":-37},"CCB11A37D49C":{"i":1,"n":"[TV] Samsung 5 Series (32)","r":-96}}}
18:37:19.696 RSL: BLE = {"Time":"2024-06-21T18:37:19","BLE":{"scans":504,"adverts":181566,"devices":2,"resets":0}}
18:42:19.639 RSL: STATE = {"Time":"2024-06-21T18:42:19","Uptime":"0T02:55:51","UptimeSec":10551,"Heap":40,"SleepMode":"Dynamic","Sleep":50,"LoadAvg":19,"MqttCount":0,"Berry":{"HeapUsed":21,"Objects":284},"Wifi":{"AP":1,"SSId":"2,4GHz RFID","BSSId":"44:4E:6D:23:7E:38","Channel":1,"Mode":"HT20","RSSI":68,"Signal":-66,"LinkCount":1,"Downtime":"0T00:00:04"}}
18:42:19.665 RSL: BLE = {"Time":"2024-06-21T18:42:19","BLEDevices":{"total":2,"F4B3B182966D":{"i":0,"n":"SBBT-002C","r":-30},"CCB11A37D49C":{"i":1,"n":"[TV] Samsung 5 Series (32)","r":-98}}}
18:42:19.688 RSL: BLE = {"Time":"2024-06-21T18:42:19","BLE":{"scans":519,"adverts":186720,"devices":2,"resets":0}}
18:47:33.380 RSL: STATE = {"Time":"2024-06-21T18:47:33","Uptime":"0T03:01:05","UptimeSec":10865,"Heap":42,"SleepMode":"Dynamic","Sleep":50,"LoadAvg":19,"MqttCount":0,"Berry":{"HeapUsed":21,"Objects":284},"Wifi":{"AP":1,"SSId":"2,4GHz RFID","BSSId":"44:4E:6D:23:7E:38","Channel":1,"Mode":"HT20","RSSI":66,"Signal":-67,"LinkCount":1,"Downtime":"0T00:00:04"}}
18:47:33.406 RSL: BLE = {"Time":"2024-06-21T18:47:33","BLEDevices":{"total":2,"F4B3B182966D":{"i":0,"n":"SBBT-002C","r":-30},"CCB11A37D49C":{"i":1,"n":"[TV] Samsung 5 Series (32)","r":-96}}}
18:47:33.419 RSL: BLE = {"Time":"2024-06-21T18:47:33","BLE":{"scans":535,"adverts":192179,"devices":2,"resets":0}}
18:52:33.367 RSL: STATE = {"Time":"2024-06-21T18:52:33","Uptime":"0T03:06:05","UptimeSec":11165,"Heap":42,"SleepMode":"Dynamic","Sleep":50,"LoadAvg":19,"MqttCount":0,"Berry":{"HeapUsed":21,"Objects":284},"Wifi":{"AP":1,"SSId":"2,4GHz RFID","BSSId":"44:4E:6D:23:7E:38","Channel":1,"Mode":"HT20","RSSI":66,"Signal":-67,"LinkCount":1,"Downtime":"0T00:00:04"}}
18:52:33.393 RSL: BLE = {"Time":"2024-06-21T18:52:33","BLEDevices":{"total":3,"F4B3B182966D":{"i":0,"n":"SBBT-002C","r":-34},"CCB11A37D49C":{"i":1,"n":"[TV] Samsung 5 Series (32)","r":-96},"88C6269DABD9":{"i":2,"r":-101}}}
18:52:33.416 RSL: BLE = {"Time":"2024-06-21T18:52:33","BLE":{"scans":550,"adverts":197376,"devices":3,"resets":0}}
18:57:33.367 RSL: STATE = {"Time":"2024-06-21T18:57:33","Uptime":"0T03:11:05","UptimeSec":11465,"Heap":39,"SleepMode":"Dynamic","Sleep":50,"LoadAvg":19,"MqttCount":0,"Berry":{"HeapUsed":21,"Objects":284},"Wifi":{"AP":1,"SSId":"2,4GHz RFID","BSSId":"44:4E:6D:23:7E:38","Channel":1,"Mode":"HT20","RSSI":68,"Signal":-66,"LinkCount":1,"Downtime":"0T00:00:04"}}
18:57:33.393 RSL: BLE = {"Time":"2024-06-21T18:57:33","BLEDevices":{"total":3,"F4B3B182966D":{"i":0,"n":"SBBT-002C","r":-36},"CCB11A37D49C":{"i":1,"n":"[TV] Samsung 5 Series (32)","r":-95},"88C6269DABD9":{"i":2,"r":-101}}}
18:57:33.417 RSL: BLE = {"Time":"2024-06-21T18:57:33","BLE":{"scans":565,"adverts":202688,"devices":3,"resets":0}}`

**Reply by sfromis** [reply]
Tasmota is seeing no beacons. As I posted, you need to "set up Tasmota to handle the messages". That means something like commands `iBeacon 1` and `BLEAddrFilter 3` (the latter value may depend on your preferences, and how the beacon works).

If/when you see the beacon messages, I'd also recommend to assigning a name `BLEAlias xxxxxxxx=Beacon`. Select your own MAC and name, of course. Then you can use `iBeaconOnlyAliased 1` to filter out other stuff which may be appearing too.

**Comment by New-three** [comment]
perfect. Shelly Blu Button is displayed as a beacon in the home menu. Alias ​​is also assigned. Many others will be found in addition.

![IMG_2067](https://github.com/arendst/Tasmota/assets/99345562/b3c24717-d71b-4e8f-a2db-062df0802e08)

Console:
0:13:04.622 RSL: SENSOR = {"Time":"2024-06-21T20:13:04","IBEACON":{"MAC":"F4B3B182966D","NAME":"Beacon","RSSI":-32,"STATE":"ON","PERSEC":0}}

**Comment by New-three** [comment]
I Try this Berry:

```
`// Define the MAC address of the beacon and the RSSI threshold
const MAC_ADDRESS = "F4B3B182966D";
const RSSI_THRESHOLD = -60;  // Note that RSSI values are typically negative

// Function to control GPIO16
def control_gpio16(state) {
    if (state) {
        gpio,16,1;  // Turn GPIO16 on
    } else {
        gpio,16,0;  // Turn GPIO16 off
    }
}

// Function to check the RSSI value and control GPIO16
def check_rssi() {
    let rssi = 0;

    // Get the RSSI value of the beacon
    let beacons = BeaconScan();

    for (let beacon in beacons) {
        if (beacon.mac == MAC_ADDRESS) {
            rssi = beacon.rssi;
            break;
        }
    }

    // Check the RSSI value and control GPIO16
    if (rssi < RSSI_THRESHOLD) {
        control_gpio16(true);  // Turn on
    } else {
        control_gpio16(false);  // Turn off
    }
}

// Main loop to periodically check the RSSI value
while (true) {
    check_rssi();
    delay(1000);  // Check every 1 second
}
`

```
I have this Problem 

![image](https://github.com/arendst/Tasmota/assets/99345562/22982e2b-9c64-40ca-a306-045aa3330515)

**Reply by sfromis** [reply]
That's because `//` is not valid syntax for comments in Berry, use `#` instead.

Your usage of blocks delimited by `{}` is not valid syntax in Berry, either. Berry is a quite different language from JavaScript or the C++ used with Arduino. Blocks are started by having a `if/for/def` etc, and ended with an `end`. You can also drop those `;` (even if they are not invalid to have).

And if you want to control a gpio, that's certainly possible, but then you need to read the docs first, instead of that weird construct you have. Simpler to just have the pin configured as a relay, instead of addressing gpios directly.

See:
https://tasmota.github.io/docs/Berry/
https://berry.readthedocs.io/en/latest/

And your `main_loop` is wholly invalid in Berry, as all functions **must** return quickly, preferably within 50 ms. You can use a `cron` timer, or maybe this small code example.
https://tasmota.github.io/docs/Berry-Cookbook/#call-function-at-intervals

Did you happen to use some AI model (ChatGPT or whatever) to get horrible advice?

Also the concept of *scanning* for RSSI values is just not workable, as a beacon is not transmitting continuously, only at intervals. If you have a beacon able to transmit at a rate satisfactory to you, you'll get all the messages rolling in anyway.

If you want to take a dive, the tool Blerry can be useful, but not so much to the very simple use case you have.
https://github.com/tony-fav/tasmota-blerry

**Reply by New-three** [reply]
Yes, I used chatgpt because I have no other option.

Would a simple script also be possible? 

Rule1  ON Beacon#F4B3B182966D(0)_RSSI<-60 DO backlog Power1 1 ENDON  ON Beacon#F4B3B182966D(0)_RSSI>-60 DO backlog Power1 0 ENDON  Rule1 1

Since I have no idea about programming, I have to use programs like chatgpt.

**Reply by sfromis** [reply]
If your best shot is having ChatGPT hallucinate, better forget it all, unless possibly if you're using *very* popular languages like Python, and have an understanding of the runtime environment you want. Tasmota is a foreign language to those LLM models which literally *knows nothing*, just picks a next work based on some guess at plausibility. For that to get useful, it needs a *massive* amount of training data first.

If you have *only* a single beacon, and nothing more, a rule set (two rules) like this might work:
```
Rule1 ON IBeacon#RSSI<-60 DO Power1 1 ENDON ON IBeacon#RSSI>-60 DO Power1 0 ENDON
Rule1 1
```

You need to match the names in the JSON accurately. And with rules you cannot look at more than one field at a time. The above will fail horribly if you have anything more arriving as `IBEACON`.

What you posted has nothing to do with a "script", but is Tasmota rules (very different thing). Triggers not working, but at least the overall syntax is in the right direction. Not specifics.

Here's a small piece of Berry code doing something like what you seem to aim for.
```
tasmota.remove_rule('IBEACON#MAC=F4B3B182966D')
tasmota.add_rule('IBEACON#MAC=F4B3B182966D',
  def (value, trigger, payload)
    var rssi = payload['IBEACON']['RSSI']
    var newstate = rssi<-60
    print(rssi, newstate, payload)
    if tasmota.get_power(0) != newstate
      tasmota.set_power(0, newstate)
    end
  end)
```
Of course, the `print` is not needed, just makes things visble. And the first `remove_rule` is only useful if you try in the Berry REPL console, to remove the old code if you want to try a variation.
For persistent install, you can have the above in the file `autoexec.be`.

The above could be shortened a bit by not doing separate variable assignments, but I left them for readability.

**Reply by New-three** [reply]
huge thanks. It works for the first time. Really amazing.

Can you tell me why the Enable Active Scan(*) item is deactivated after every boot?

Would you be willing to customize the Berry for me in the future for a donation?

I will start the setup now and see if everything works as I hope.

**Reply by sfromis** [reply]
In general, you should not expect any of the BLE options you select to be persistent across boots, but have some sort of automatic reissuing of commands. Simplest is probably to put them in the file `autoexec.bat`, one command per line.

And no, I'm not interested in providing support on request, and especially not when it gets to be some sort of "paid". In general, with open source and Tasmota, the concept is that you can ask questions. Someone *may* have something to add. Or not. It generally helps if you show that you make an effort, but stumble at some point.

I posted the Berry code example as a sort of "getting started", and I already posted links about where to learn more about Berry.

**Reply by New-three** [reply]
I created an autoexec.bat and entered two commands like BLEscan and IBeacon. It actually works :) Can I insert the Berry code here too? It is no longer there after every restart.

**Reply by sfromis** [reply]
No. I already said that you can have it in `autoexec.be` instead.
